### DISCUSSION TITLE
global Livewire object not available in time

### DISCUSSION THREAD

**Discussion by finn-matti** [livewire/livewire#7227] [1. Help] [answered]
I added livewire to my laravel project and have noticed that I can't work with the global Livewire object in a script tag because the lifewire scripts aren't loaded in time.

>By default, Livewire injects the JavaScript and CSS assets it needs into each page that includes a Livewire component.

According to [this](https://livewire.laravel.com/docs/installation#manually-including-livewires-frontend-assets) I shouldn't be required to include `@livewireStyles` and `@livewireScripts` by hand into my app.blade.php layout, but I can't seem to make it work reliably otherwise.

This is how my app.blade.php looks:

```html
<!DOCTYPE html>
<html lang="{{ str_replace('_', '-', app()->getLocale()) }}">
    <head>
        <meta charset="utf-8">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <meta name="csrf-token" content="{{ csrf_token() }}">

        <title>{{ config('app.name', 'Laravel') }}</title>

        <!-- Fonts -->
        <link rel="preconnect" href="https://fonts.bunny.net">
        <link href="https://fonts.bunny.net/css?family=figtree:400,500,600&display=swap" rel="stylesheet" />

        <!-- Scripts -->
        @vite(['resources/css/app.css', 'resources/js/app.js'])
        @livewireStyles
    </head>
    <body class="font-sans antialiased">
        <!-- FIXME: I don't like that this is here. But otherwise the Livewire global object is not available in time it seems -->
        @livewireScripts
        <div class="min-h-screen bg-gray-100 dark:bg-gray-900">
            @include('layouts.navigation')

            <!-- Page Heading -->
            @if (isset($header))
                <header class="bg-white dark:bg-gray-800 shadow">
                    <div class="max-w-7xl mx-auto py-6 px-4 sm:px-6 lg:px-8">
                        {{ $header }}
                    </div>
                </header>
            @endif

            <!-- Page Content -->
            <main>
                {{ $slot }}
            </main>
        </div>
    </body>
</html>
```

And this is a dashboard.blade.php that uses a component and the global Livewire object in js:

```html
<x-app-layout>
    <x-slot name="header">
        <h2 class="font-semibold text-xl text-gray-800 dark:text-gray-200 leading-tight">
            {{ __('Dashboard') }}
        </h2>
    </x-slot>

    <div class="py-12">
        <div class="max-w-7xl mx-auto sm:px-6 lg:px-8">
            <div class="bg-white dark:bg-gray-800 overflow-hidden shadow-sm sm:rounded-lg">
                <div class="p-6 text-gray-900 dark:text-gray-100">
                    {{ __("You're logged in!") }}
                </div>
            </div>
        </div>
    </div>

    @livewire('delete-markdown-files')

    <script>
        Livewire.on('alert', data => {
            // data is an array with one element
            alert(data[0].message);
        });
    </script>
</x-app-layout>
```

Without the manual addition of @livewireScripts I get the following error in the console of my browser:

```
ReferenceError: Can't find variable: Livewire
```

But this only happens sometimes (which I assume is a cashing issue/race condition).

Am I interpreting how to do this wrong? (e.g. I should be able to access the Livewire object without any special incantations) I would imagine that I shouldn't need to put in too much manual work/configuration into this since it should work out of the box, no?

**Comment by joshhanley** [comment, marked as answer]
@finn-matti you need to wrap any of your code that needs Livewire in a `livewire:init` document listener. See here for details https://livewire.laravel.com/docs/javascript#global-livewire-events

Hope this helps!

**Reply by finn-matti** [reply]
It did! Somehow I missed that part. Thanks!

**Reply by joshhanley** [reply]
No worries! 🙂
